src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+			break;
+		case BITS_0_1:
+			m_memcache_seq = event == 7 ? BITS_0_1_7 : (event == 6 ? BITS_6 : BITS_ERR);
+			break;
+		case BITS_6:
+			m_memcache_seq = event == 7 ? BITS_6_7 : (event == 0 ? BITS_0 : BITS_ERR);
+			break;
+		case BITS_7:
+			m_memcache_seq = event == 1 ? BITS_7_1 : BITS_ERR;
+			break;
+		case BITS_EMPTY:
+			m_memcache_seq = event == 7 ? BITS_7 : (event == 6 ? BITS_6 : (event == 0 ? BITS_0 : BITS_ERR));
+			break;
+		default:
+			m_memcache_seq = BITS_ERR;
+	}
+
+	LOG("memcache seq = %08x\n", m_memcache_seq);
+}
+
+void sega_9h0_0008_state::memcache_parse_data_bit(uint32_t &status)
+{
+	if (m_memcache_seq == BITS_0_1) {
+		// After a command has been ack'd, we might reset instead of parse the command's data bits.
+		// Due to both sequences starting with the same prefix, we have to bailout on the shorter
+		// sequence if the lookahead matches the longer sequence.
+		if (!m_memcache_events.empty() && m_memcache_events.front() == 7) {
+			LOG("memcache READ_COMMAND 0 1 skipped\n");
+			return;
+		}
+		m_memcache_seq = BITS_EMPTY;
+		m_memcache_data = (m_memcache_data << 1) | 0;
+		m_memcache_i++;
+	} else if (m_memcache_seq == BITS_6_7) {
+		m_memcache_seq = BITS_EMPTY;
+		m_memcache_data = (m_memcache_data << 1) | 1;
+		m_memcache_i++;
+	} else if (m_memcache_seq == BITS_0_1_7) {
+		status = 1;
+		m_memcache_seq = BITS_EMPTY;
+		m_memcache_state = IDLE;
+		m_memcache_state_after = {};
+		LOG("memcache READ_COMMAND 0 1 7 -> IDLE\n");
+	}
+}
+
+uint32_t sega_9h0_0008_state::io_expansion_r()
+{
+	return 0; // TODO
+}
+
+uint32_t sega_9h0_0008_state::io_memcache_r()
+{
+	/*
+	    A non-memory mapped cache can be interfaced with data address 0x500300c0,
+	    while the status after parsing this data (events) can be read at 0x500300bc.
+
+	    The cache stores RAM values in range 0xc00c6400..0xc00c6fff.
+
+	    During I/O init, this cache is cleared.
+
+	    During Test Mode, some memory integrity checks are carried out,
+	    consisting of reading RAM to the cache, then writing those cached bytes
+	    at the same RAM address, which is then compared with the expected values.
+
+	    The communication protocol consists of sending sequences of events,
+	    contextual to the command that is currently being parsed (which in turn
+	    is initiated with a sequence of events).
+
+	    Simplified flow:
+
+	        IDLE
+	        READ_COMMAND (includes RAM address bits 8..12)
+	            if 0xa0:
+	                READ_ADDRESS_BYTE (RAM address bits 0..7)
+	                READ_RAM_BYTE (RAM value bits 0..7)
+	            if 0xa1:
+	                WRITE_RAM_BYTE (assumed to be at previous read RAM address)
+
+	    Some of these sequences can be followed by acknowledgements (READ_ACK), to
+	    ensure sent events were parsed as expected, otherwise games bailout from the
+	    session. Always ends with the same sequence (BITS_0_1_7).
+	*/
+	uint32_t status = 0;
+	while (!m_memcache_events.empty()) {
+		memcache_advance(status);
```

Comment:
This is another read handler with side effects – you need to check `machine().side_effects_disabled()`.